README.md
# NAME

gitgitr - Automatically fetch and build the most recent git.

# VERSION

version 0.6

# SYNOPSIS

    gitgitr
      -- fetches and builds the most recent version of git

    gitgitr -t
      -- fetches and builds the most recent version of git, and runs the
         test suite prior to installation

    gitgitr -v 1.8.0
      -- fetches and builds version 1.8.0 of git

# DESCRIPTION

`gitgitr` is a tiny utility to simplify building the most recent (or,
really, any arbitrary) version of `git`. This is something you
probably only really need if you're obsessive about running the most
recent version of `git`, or if you maintain something like
[Git::Wrapper](https://metacpan.org/pod/Git::Wrapper), where the ability to quickly install a particular
`git` version comes up way more often than you would like.

# OPTIONS

- --help / -h

    Displays basically the same information you're currently soaking in.

- --no\_symlink / -N

    Don't symlink `/opt/git` to the new build directory

- --prefix / -p

    Installation prefix for the current build. Defaults to
    `/opt/git-$GIT_VERSION_NUMBER`

- --reinstall / -r

    By default, `gitgitr` will not build a version that's already
    installed. This flag will force it to build a version even if it is
    already preset.

- --run\_tests / -t

    Makes `gitgitr` run the `git` test suite after building and before
    installing. If the test suite fails, the build will not be
    installed. Disabled by default because the test suite takes a really
    long time to run.

- --verbose / -V

    Makes `gitgitr` be a lot more chatty about what it's doing

- --version / -v

    Specifies what version of `git` to build. Defaults to the most recent
    version as found on the front page of [http://git-scm.com](http://git-scm.com)

# SEE ALSO

- [http://git-scm.com](http://git-scm.com)

# AUTHOR

John SJ Anderson <genehack@genehack.org>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by John SJ Anderson.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Changes file for {{$dist->name}}

{{$NEXT}}

0.907     2020-04-20 15:55:18-07:00 America/Los_Angeles

 Update repo metadata

0.906     2017-03-20 15:14:59-07:00 America/Los_Angeles

  Switch the _other_ LWP::Simple function call to HTTP::Tiny too...

0.905     2017-03-20 15:03:15-07:00 America/Los_Angeles

  Switch from LWP::Simple to HTTP::Tiny

0.904     2017-03-20 15:01:45-07:00 America/Los_Angeles

  Update copyright notice

0.903     2016-02-15 09:59:11-08:00 America/Los_Angeles

  Don't use OpenSSL on MacOS X

0.902     2015-06-27 13:53:01-07:00 America/Los_Angeles

  Move App::GitGitr::Cmd to App::GitGitr. Should make PAUSE happier.

0.901     2015-06-27 13:23:07-07:00 America/Los_Angeles

        Doc tweak
        Extra error checking
        Remove FindBin stuff

0.900     2015-04-14 21:14:43-07:00 America/Los_Angeles

        No changes since last dev release.

0.8       2015-03-29 23:54:53-04:00 America/New_York (TRIAL RELEASE)

        Add compilation tests

0.7       2015-03-28 21:02:10-07:00 America/Los_Angeles (TRIAL RELEASE)

0.6       2014-08-16 11:21:34-07:00 America/Los_Angeles

        CPAN DAY RELEASE!

0.5       2014-05-31 11:41:30-07:00 America/Los_Angeles

        Doc tweaks
        Switch homepage to pretty version
        Use a real HTML parser to extract version
        Switch back to kernel.org for downloads

0.4       2014-04-06 12:20:14-07:00 America/Los_Angeles

        Added 'no_symlink' option
        General cleanup & stuff
        Documentation (thanks to Tom Metro for the asskick)
        Update dist.ini to generate Makefile.PL in repo
        Update dist.ini to generate README.md in build & repo

0.3       2014-04-05 23:41:32-07:00 America/Los_Angeles

        Dummy App::GitGitr package

0.2       2014-04-05 23:33:07-07:00 America/Los_Angeles

        Created Changes file


lib/App/GitGitr.pm
package App::GitGitr;

# ABSTRACT: GitGitr command support. See L<gitgitr> for full documentation.

use parent 'App::Cmd::Simple';

use strict;
use warnings;
use 5.010;

use autodie qw/ :all /;

use Archive::Extract;
use Carp;
use File::Remove 'remove';
use HTML::TreeBuilder::XPath;
use HTTP::Tiny;

sub opt_spec {
  return (
    [ "no_symlink|N" => 'do not symlink final build to /opt/git' ],
    [ "prefix|p"     => 'directory to install under (defaults to /opt/git-$VERSION)' ],
    [ "reinstall|r"  => 'build even if installation directory exists (default=false)' ],
    [ "run_tests|t"  => 'run "make test" after building' ] ,
    [ "verbose|V"    => 'be verbose about what is being done' ] ,
    [ "version|v=s"  => 'Which git version to build. Default = most recent' ] ,
  );
}

sub execute {
  my( $self , $opt , $args ) = @_;

  my $version     = $opt->{version} // _build_version();
  my $install_dir = $opt->{prefix}  // "/opt/git-$version";

  say "CURRENT VERSION: $version"
    if $opt->{verbose};

  if ( -e $install_dir and ! $opt->{reinstall} ) {
    if( $opt->{no_symlink} ) {
      say "Most recent version ($version) already installed.";
    }
    else {
      $self->_symlink( $opt , $version );
      say "Most recent version ($version) already installed. /opt/git redirected to that version";
    }
  }
  else {
    chdir( '/tmp' );

    say "BUILD/INSTALL git-$version"
      if $opt->{verbose};

    my $pkg_path = $self->_download( $opt , $version );
    $self->_extract( $opt , $pkg_path );
    $self->_configure( $opt , $version , $install_dir );
    $self->_make( $opt );
    $self->_make_test( $opt ) if $opt->{run_tests};
    $self->_make_install( $opt );
    $self->_cleanup( $opt , $version );
    $self->_symlink( $opt , $version ) unless $opt->{no_symlink};

    say "\n\nBuilt new git $version."
      if $opt->{verbose};
    say "/opt/git symlink switched to new version ($version)."
      unless $opt->{no_symlink};
  }

  die "No new version?!"
    unless -e "/opt/git-$version";

}

sub _build_version {
  my $response = HTTP::Tiny->new->get('http://git-scm.com/');
  unless ( $response->{success} ) {
    my $error = sprintf "Failed to fetch content from Git web page!\nSTATUS: %s REASON: %s" ,
      $response->{status}, $response->{reason};
    croak $error;
  }

  my $tree = HTML::TreeBuilder::XPath->new;
  $tree->parse_content( $response->{content} )
    or croak "Failed to parse content from Git web page!";

  my $version = $tree->findvalue('/html/body//span[@class="version"]')
    or croak "Can't parse version from Git web page!";
  $version =~ s/^\s*//; $version =~ s/\s*$//;

  return $version;
}

sub _download {
  my( $self , $opt , $version ) = @_;

  say "*** download" if $opt->{verbose};

  my $pkg_path = sprintf "git-%s.tar.gz" , $version;

  my $url = sprintf "https://kernel.org/pub/software/scm/git/%s" , $pkg_path;
  #my $url = sprintf "http://git-core.googlecode.com/files/%s" , $pkg_path;

  my $response = HTTP::Tiny->new->mirror( $url , $pkg_path );
  unless ( $response->{success} ) {
    my $message = sprintf "DOWNLOAD FAILED!\n** STATUS: %s REASON: %s",
      $response->{status}, $response->{reason};
    croak $message;
  }

  return $pkg_path;
};

sub _extract {
  my( $self , $opt , $pkg_path ) = @_;
  say "*** extract" if $opt->{verbose};
  my $ae = Archive::Extract->new( archive => $pkg_path );
  $ae->extract or die $ae->error;
  unlink $pkg_path;
};

sub _configure {
  my( $self , $opt , $version , $install_dir ) = @_;
  say "*** configure" if $opt->{verbose};
  chdir "git-$version";
  ### FIXME should have some way to allow override of these args
  my $cmd = "./configure --prefix=$install_dir --without-tcltk";
  # MacOS doesn't have openssl.h anymore, i guess?
  $cmd   .= " --without-openssl" if $^O eq 'darwin';
  _run( $cmd );
};

sub _make {
  my( $self , $opt ) = @_;
  say "*** make" if $opt->{verbose};
  _run( 'make' );
};

sub _make_test {
  my( $self , $opt ) = @_;
  say "*** make test" if $opt->{verbose};
  _run( 'make test' );
};

sub _make_install {
  my( $self , $opt ) = @_;
  say "*** make install" if $opt->{verbose};
  _run( 'make install' );
};

sub _cleanup {
  my( $self , $opt , $version ) = @_;
  say "*** cleanup" if $opt->{verbose};
  chdir '..';
  remove( \1 , "git-$version" );
};

sub _symlink {
  my( $self , $opt , $version ) = @_;
  say "*** symlink" if $opt->{verbose};
  chdir '/opt';
  remove( 'git' );
  symlink( "git-$version" , 'git' );
};

sub _run {
  my $arg = shift;
  $arg .= ' 2>&1 >/dev/null';
  system( $arg ) == 0;
}

1;


lib/App/GitGitr/Cmd.pm
package App::GitGitr::Cmd;

# ABSTRACT: GitGitr command support. See L<gitgitr> for full documentation.

use parent 'App::Cmd::Simple';

use strict;
use warnings;
use 5.010;

use autodie qw/ :all /;

use Archive::Extract;
use Carp;
use File::Remove 'remove';
use HTML::TreeBuilder::XPath;
use LWP::Simple;

sub opt_spec {
  return (
    [ "no_symlink|N" => 'do not symlink final build to /opt/git' ],
    [ "prefix|p"     => 'directory to install under (defaults to /opt/git-$VERSION)' ],
    [ "reinstall|r"  => 'build even if installation directory exists (default=false)' ],
    [ "run_tests|t"  => 'run "make test" after building' ] ,
    [ "verbose|V"    => 'be verbose about what is being done' ] ,
    [ "version|v=s"  => 'Which git version to build. Default = most recent' ] ,
  );
}

sub execute {
  my( $self , $opt , $args ) = @_;

  my $version     = $opt->{version} // _build_version();
  my $install_dir = $opt->{prefix}  // "/opt/git-$version";

  say "CURRENT VERSION: $version"
    if $opt->{verbose};

  if ( -e $install_dir and ! $opt->{reinstall} ) {
      if( $opt->{no_symlink} ) {
          say "Most recent version ($version) already installed.";
      }
      else {
        $self->_symlink( $opt , $version );
        say "Most recent version ($version) already installed. /opt/git redirected to that version";
      }
  }
  else {
    chdir( '/tmp' );

    say "BUILD/INSTALL git-$version"
      if $opt->{verbose};

    my $pkg_path = $self->_download( $opt , $version );
    $self->_extract( $opt , $pkg_path );
    $self->_configure( $opt , $version , $install_dir );
    $self->_make( $opt );
    $self->_make_test( $opt ) if $opt->{run_tests};
    $self->_make_install( $opt );
    $self->_cleanup( $opt , $version );
    $self->_symlink( $opt , $version ) unless $opt->{no_symlink};

    say "\n\nBuilt new git $version."
      if $opt->{verbose};
    say "/opt/git symlink switched to new version ($version)."
      unless $opt->{no_symlink};
  }

  die "No new version?!"
    unless -e "/opt/git-$version";

}

sub _build_version {
  my $content = get( 'http://git-scm.com/' );
  my $tree = HTML::TreeBuilder::XPath->new;
  $tree->parse_content( $content )
    or croak "Failed to parse content from Git web page!";
  my $version = $tree->findvalue('/html/body//span[@class="version"]')
    or croak "Can't parse version from Git web page! $content";
  $version =~ s/^\s*//; $version =~ s/\s*$//;
  return $version;
}

sub _download {
  my( $self , $opt , $version ) = @_;
  say "*** download" if $opt->{verbose};
  my $pkg_path = sprintf "git-%s.tar.gz" , $version;
  my $url = sprintf "https://kernel.org/pub/software/scm/git/%s" , $pkg_path;
  #my $url = sprintf "http://git-core.googlecode.com/files/%s" , $pkg_path;
  my $ret = getstore( $url , $pkg_path );
  die $ret unless $ret eq '200';
  return $pkg_path;
};

sub _extract {
  my( $self , $opt , $pkg_path ) = @_;
  say "*** extract" if $opt->{verbose};
  my $ae = Archive::Extract->new( archive => $pkg_path );
  $ae->extract or die $ae->error;
  unlink $pkg_path;
};

sub _configure {
  my( $self , $opt , $version , $install_dir ) = @_;
  say "*** configure" if $opt->{verbose};
  chdir "git-$version";
  ### FIXME should have some way to allow override of these args
  _run( "./configure --prefix=$install_dir --without-tcltk" );
};

sub _make {
  my( $self , $opt ) = @_;
  say "*** make" if $opt->{verbose};
  _run( 'make' );
};

sub _make_test {
  my( $self , $opt ) = @_;
  say "*** make test" if $opt->{verbose};
  _run( 'make test' );
};

sub _make_install {
  my( $self , $opt ) = @_;
  say "*** make install" if $opt->{verbose};
  _run( 'make install' );
};

sub _cleanup {
  my( $self , $opt , $version ) = @_;
  say "*** cleanup" if $opt->{verbose};
  chdir '..';
  remove( \1 , "git-$version" );
};

sub _symlink {
  my( $self , $opt , $version ) = @_;
  say "*** symlink" if $opt->{verbose};
  chdir '/opt';
  remove( 'git' );
  symlink( "git-$version" , 'git' );
};

sub _run {
  my $arg = shift;
  $arg .= ' 2>&1 >/dev/null';
  system( $arg ) == 0;
}

1;


lib/App/Gitgitr/Cmd/Command/build.pm
use MooseX::Declare;
class App::Gitgitr::Cmd::Command::build extends MooseX::App::Cmd::Command {
  use Archive::Extract;
  use Carp;
  use File::Remove 'remove';
  use LWP::Simple;
  use 5.010;

  has run_tests => (
    isa           => 'Bool' ,
    is            => 'rw' ,
    cmd_aliases   => 't',
    documentation => 'run "make test" after building',
    traits        => [ qw(Getopt) ] ,
    default       => 0 ,
  );

  has verbose => (
    isa           => 'Bool' ,
    is            => 'rw' ,
    cmd_aliases   => 'V' ,
    documentation => 'be verbose about what is being done' ,
    traits        => [ qw(Getopt) ] ,
    default       => 0 ,
  );

  has version => (
    isa           => 'Str' ,
    is            => 'rw'  ,
    cmd_aliases   => 'v' ,
    documentation => 'Which git version to download and build. Defaults to most current.' ,
    lazy_build    => 1 ,
    traits        => [ qw(Getopt)] ,
  );

  method _build_version {
    my $content = get( 'http://git-scm.com/' );
    my( $version ) = $content =~ m|<div id="ver">v([\d\.]+)</div>|
      or croak "Can't parse version from Git web page!";

    return $version;
  };

  has _pkg_path => (
    isa    => 'Str' ,
    is     => 'rw' ,
    traits => [ qw(NoGetopt)] ,
  );

  method execute {
    my $version = $self->version;

    say "CURRENT VERSION: $version"
      if $self->verbose;

    unless ( -e "/opt/git-$version" ) {
      chdir( '/tmp' )
        or die "Can't cd to /tmp";

      say "BUILD/INSTALL git-$version"
        if $self->verbose;

      $self->_download();
      $self->_extract();
      $self->_configure();
      $self->_make();
      $self->_make_test() if $self->run_tests;
      $self->_make_install();
      $self->_cleanup();

      say "\n\nBuilt new git $version."
        if $self->verbose;
    }

    die "No new version?!"
      unless -e "/opt/git-$version";

    $self->_symlink();

    say "New version ($version) symlinked into /opt/git";
  };

  method _download {
    say "*** download" if $self->verbose;
    $self->_pkg_path( sprintf "git-%s.tar.gz" , $self->version );
    my $url = sprintf "http://kernel.org/pub/software/scm/git/%s" , $self->_pkg_path;
    my $ret = getstore( $url , $self->_pkg_path );
    die $ret unless $ret eq '200';
  };

  method _extract {
    say "*** extract" if $self->verbose;
    my $ae = Archive::Extract->new( archive => $self->_pkg_path );
    $ae->extract or die $ae->error;
    unlink $self->_pkg_path;
  };

  method _configure {
    say "*** configure" if $self->verbose;
    my $version = $self->version;
    chdir "git-$version";
    $self->_run( "./configure --prefix=/opt/git-$version" );
  };

  method _make {
    say "*** make" if $self->verbose;
    $self->_run( 'make' );
  };

  method _make_test {
    say "*** make test" if $self->verbose;
    $self->_run( 'make test' );
  };

  method _make_install {
    say "*** make install" if $self->verbose;
    $self->_run( 'make install' );
  };

  method _cleanup {
    say "*** cleanup" if $self->verbose;
    my $version = $self->version;
    chdir '..';
    remove( \1 , "git-$version" );
  };

  method _symlink {
    say "*** symlink" if $self->verbose;
    my $version = $self->version;
    chdir '/opt';
    remove( 'git' );
    symlink( "git-$version" , 'git' );
  };

  method _run ( $arg ) {
    $arg .= ' 2>&1 >/dev/null';
    system( $arg ) == 0
      or die "$arg failed ($?)";
  };

}

__END__

=head1 NAME

App::Gitgitr::Cmd::Command::build - Fetch a git version and build it.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Automatically fetch and build the most recent git.",
  "AUTHOR" => "John SJ Anderson <genehack\@genehack.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-GitGitr",
  "EXE_FILES" => [
    "bin/gitgitr"
  ],
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.010",
  "NAME" => "App::GitGitr",
  "PREREQ_PM" => {
    "App::Cmd::Simple" => 0,
    "Archive::Extract" => 0,
    "Carp" => 0,
    "File::Remove" => 0,
    "HTML::TreeBuilder::XPath" => 0,
    "HTTP::Tiny" => 0,
    "IPC::System::Simple" => 0,
    "autodie" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0,
    "blib" => "1.01"
  },
  "VERSION" => "0.905",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "App::Cmd::Simple" => 0,
  "Archive::Extract" => 0,
  "Carp" => 0,
  "File::Remove" => 0,
  "File::Spec" => 0,
  "HTML::TreeBuilder::XPath" => 0,
  "HTTP::Tiny" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "IPC::System::Simple" => 0,
  "Test::More" => 0,
  "autodie" => 0,
  "blib" => "1.01",
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
